### DISCUSSION TITLE
Copying multiple elements in Builder

### DISCUSSION THREAD

**Discussion by jjriggs** [ImpulseAdventure/GUIslice#355] [Q&A] [answered]
Is it possible to duplicate an entire page and the elements it contains in Builder? I find it's very cumbersome and time consuming to copy and paste individual elements when creating a new page. Many GUI pages in my project are very similar to one another, with Cancel/OK buttons, for example, always in the same place. It would save a ton of time if I could just duplicate an entire page at once. This would also be fairly easy to do outside the Builder environment if the .prj files were some sort of plaintext (JSON or XML, for example) format, but since they're a binary, this isn't possible.
Sidenote: it's also very annoying that copy/paste does not preserve the X/Y coordinates of the original element.

Jordan

**Comment by Pconti31** [comment, marked as answer]
@jjriggs The answer is yes you can copy full pages or any number of Elements at once.   Easy if you use Page layout, Rectangle Selection tool.
The Builder can't preserve the X/Y coordinates on the same page since you wouldn't notice the copy. 

However, If you go to a New Page or a different page the Builder does preserve the X/Y coordinates.

You can even copy Elements between projects as long as you don't close the builder since the elements are copied to the Java internal clipboard.
Paul-- 
Author of the Builder.

**Comment by Pconti31** [comment]
Actually the algorithm for copy is If we are only pasting one widget work out an offset for x,y position so  
it doesn't get placed on top of the original.  More than one widget on the clipboard assume copying to another page
so keep original x and y positions.
Paul--

**Comment by jjriggs** [comment]
Paul,

OK, thanks for the tip. I was getting tripped up because I was trying to select multiple elements from the tree view in the left panel, which you cannot do. It would be great if in a future version, one could select the parent page element and copy it and its child elements all at once.
I understand your logic behind the X/Y coordinates. I must be misremembering, because I recall having things jump around on me when I copied from one page into another. But now that I've tried it again, it works you described.
One other thing that's been a bit of an issue is the rectangular selection tool. For some reason, the box expands from some place off-canvas, rather than where I clicked the mouse, so it's very difficult to use. Is that how it's supposed to work, or did I just hit a glitch?

Thanks,
Jordan

**Comment by Pconti31** [comment]
@jjriggs I would guess you have set your TFT display screen incorrectly or that you are not showing the full size of TFT Simulation screen.

First check under Project Options, your settings for TFT Screen Width and TFT Screen Height.  

If they are correct, be sure you have sized your TFT Simulation panel to show all of your display or it will place things off screen.

Lastly, be sure your are using the correct version of Java.  Arduino IDE uses Java 8 (1.8) as do we.  Java 14 and 15 cause all sorts of bad behavior.   See the wiki how to install the correct versions of Java.
Paul--

**Comment by Pconti31** [comment]
@jjriggs Actually as I look over the code I don't see how the cursor could be placed off the TFT Simulation.
All the tool does is change the cursor to a Cross and wait for you to do the left press and drag.  No click involved.
If you instead left click the cursor is reset to a Pointer and the Rectangle Tool is disabled and you need to select it again.
Paul--
